server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -4554,6 +4571,71 @@ public void finalizeStop(VirtualMachineProfile profile, Answer answer) {
         return vmParamPair;
     }
 
+    private void validateDeploymentDestination(Pod destinationPod, Cluster destinationCluster, Host destinationHost) {
+
+        if (destinationHost != null) {
+            if (destinationCluster != null) {
+                if (destinationHost.getClusterId() != destinationCluster.getId()) {
+                    throw new InvalidParameterValueException("Host " + destinationHost.getId() + " is not a child of provided cluster " + destinationCluster.getId());
+                }
+                if (destinationPod != null) {
+                    if (destinationCluster.getPodId() != destinationPod.getId()) {
+                        throw new InvalidParameterValueException("Cluster " + destinationCluster.getId() + " is not a child of provided pod " + destinationPod.getId());
+                    }
+                }
+            }
+        }
+    }
+
+    private Pod getDestinationPod(Long podId, boolean isRootAdmin) {
+
+        Pod destinationPod = null;
+        if (podId != null) {
+            if (!isRootAdmin) {
+                throw new PermissionDeniedException(
+                        "Parameter " + ApiConstants.POD_ID + " can only be specified by a Root Admin, permission denied");
+            }
+            destinationPod = _podDao.findById(podId);
+            if (destinationPod == null) {
+                throw new InvalidParameterValueException("Unable to find the pod to deploy the VM, pod id=" + podId);
+            }
+        }
+        return destinationPod;
+    }
+
+    private Cluster getDestinationCluster(Long clusterId, boolean isRootAdmin) {
+
+        Cluster destinationCluster = null;
+        if (clusterId != null) {
+            if (!isRootAdmin) {
```

Comment:
so a user can not specify a cluster they have acces to?

Reply:
This feature is only available to Admin users. (Root and Domain)